### PR TITLE
HC control of default boundary condition - part two

### PR DESCRIPTION
**Pull request by jbathmann**
According to the discussion in issue #2414 a new feature is added to the HC-implementation:
An additional, optional tag in the project file is added to decide whether HC is assembled in the advective or the conventional form. However, by default HC is assembled in the advective form. 

Hereby, this PR aims to allow the user to choose default boundary conditions according to its needs.

### CODE REVIEW

**Comment by endJunction** [comment]
I'd recommend to separate changes of HC-FEM from the new boundary condition.

**Comment by jbathmann** [comment]
> I'd recommend to separate changes of HC-FEM from the new boundary condition.

PR #2467 contains the first part:
-changes of HC-FEM
-test updates

**Comment by jbathmann** [comment]
Just updated this PR such that it contains the 2nd part of PR #2467. 
-New boundary condition allowing free component flow in non_advective form of hc.
-Test to prove mass conservation i.e. impermeability of a boundary even though there is a nonzero bulk flow across the boundary
-Test of the new boundary condition by mapping the non advective form to the advective form with open boundaries

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        bc_mesh.getProperties().template getPropertyVector<std::size_t>(
+            "bulk_node_ids", MeshLib::MeshItemType::Node, 1);
+
+    std::vector<MeshLib::Element*> const& boundary_elements =
+        bc_mesh.getElements();
+    std::vector<MeshLib::Element*> const& bulk_elements =
+        bulk_mesh.getElements();
+    std::vector<std::size_t> bulk_element_ids;
+    std::vector<std::size_t> bulk_face_ids;
+    if (!(global_dim == 3))
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp]: Minor: `!(global_dim == 3)` is equivalent to `global_dim != 3` which is probably easier to read.

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        OGS_FATAL(
+            "HCNonAdvectiveFreeComponentFlowBoundary is only implemented for "
+            "2D boundary meshes");
+    }
+    unsigned node_bulk_mesh_id(0);
+    unsigned node_boundary_mesh_id(0);
+    for (auto const* bulk_element : bulk_elements)
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp]: I don't understand this for loop. Why don't you loop only over the boundary elements at first? Each boundary element has already the link to its bulk element.

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        HCNonAdvectiveFreeComponentFlowBoundaryConditionData const& data)
+        : Base(e, is_axially_symmetric, integration_order),
+          _data(data),
+          _local_matrix_size(local_matrix_size),
+          _bulk_element_id(data.bulk_element_ids[e.getID()]),
+          _bulk_face_id(data.bulk_face_ids[e.getID()]),
+          _surface_eigenvector(giveEigenvector(e))
+    {
+    }
+    Eigen::RowVector3d giveEigenvector(MeshLib::Element const& e)
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]: The name of the method is rather generic. Is my understanding correct that the method returns the surface normal? Then the name could also reflect the nature of the returned vector. Maybe `getSurfaceNormal()`?

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp (hunk 1 above)
Comment (jbathmann) [reply ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp]: Changed.

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h (hunk 3 above)
Comment (jbathmann) [reply ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]: Changed

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp (hunk 2 above)
Comment (jbathmann) [reply ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp]: I changed it such that we assume the identification of bulk_element_- and bulk_face_ids is already done in the preprocessing. However, I would suggest to change the identifysubdomains tool accordingly. Currently it is just storing the bulk_element_ids but does not consider the face_ids.

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+{
+    DBUG(
+        "Constructing open boundary for Component Transport process from "
+        "config.");
+    //! \ogs_file_param{prj__process_variables__process_variable__boundary_conditions__boundary_condition__type}
+    config.checkConfigParameter("type",
+                                "HCNonAdvectiveFreeComponentFlowBoundary");
+
+    auto const boundary_permeability_name =
+        //! \ogs_file_param{prj__process_variables__process_variable__boundary_conditions__boundary_condition__VariableDependentNeumann__constant_name}
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryCondition.cpp]:
```suggestion
        //! \ogs_file_param{prj__process_variables__process_variable__boundary_conditions__boundary_condition__HCNonAdvectiveFreeComponentFlowBoundary__parameter}
```

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+#include "ParameterLib/MeshNodeParameter.h"
+#include "ProcessLib/Process.h"
+
+namespace ProcessLib
+{
+struct HCNonAdvectiveFreeComponentFlowBoundaryConditionData
+{
+    ParameterLib::Parameter<double> const& boundary_permeability;
+    MeshLib::PropertyVector<std::size_t> const bulk_face_ids;
+    MeshLib::PropertyVector<std::size_t> const bulk_element_ids;
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]: Not sure at this point (of review) but is copying the property vectors intended? I'd guess a const-ref is sufficient.

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        bool const is_axially_symmetric,
+        unsigned const integration_order,
+        HCNonAdvectiveFreeComponentFlowBoundaryConditionData const& data)
+        : Base(e, is_axially_symmetric, integration_order),
+          _data(data),
+          _local_matrix_size(local_matrix_size),
+          _surface_eigenvector(getOrientedSurfaceNormal(e))
+    {
+    }
+    Eigen::RowVector3d getOrientedSurfaceNormal(MeshLib::Element const& e)
```
Comment (endJunction) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]: Why not `MeshLib::FaceRule::getSurfaceNormal`?

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        NodalVectorType const boundary_permeability_node_values =
+            _data.boundary_permeability.getNodalValuesOnElement(Base::_element,
+                                                                t);
+        unsigned const n_integration_points =
+            Base::_integration_method.getNumberOfPoints();
+
+        auto const indices =
+            NumLib::getIndices(mesh_item_id, dof_table_boundary);
+        std::vector<double> const local_values = x.get(indices);
+        std::size_t bulk_element_id =
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]:
```suggestion
        std::size_t const bulk_element_id =
```

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            _data.bulk_element_ids[Base::_element.getID()];
+        std::size_t bulk_face_id = _data.bulk_face_ids[Base::_element.getID()];
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]:
```suggestion
        std::size_t const bulk_face_id = _data.bulk_face_ids[Base::_element.getID()];
```

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler(
+        MeshLib::Element const& e,
+        std::size_t const local_matrix_size,
+        bool const is_axially_symmetric,
+        unsigned const integration_order,
+        HCNonAdvectiveFreeComponentFlowBoundaryConditionData const& data)
+        : Base(e, is_axially_symmetric, integration_order),
+          _data(data),
+          _local_matrix_size(local_matrix_size),
+          _surface_eigenvector(getOrientedSurfaceNormal(e))
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]:
```suggestion
          _surface_normal(getOrientedSurfaceNormal(e))
```

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            auto const bulk_element_point = MeshLib::getBulkElementPoint(
+                _data.process.getMesh(), bulk_element_id, bulk_face_id, wp);
+
+            double int_pt_value = 0.0;
+            NumLib::shapeFunctionInterpolate(local_values, N, int_pt_value);
+
+            NodalVectorType const neumann_node_values =
+                -boundary_permeability_node_values * int_pt_value *
+                _data.process.getFlux(bulk_element_id, bulk_element_point, t, x)
+                    .dot(_surface_eigenvector);
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]:
```suggestion
                    .dot(_surface_normal);
```

File: ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            _local_rhs.noalias() += N * neumann_node_values.dot(N) * w;
+        }
+
+        b.add(indices, _local_rhs);
+    }
+
+private:
+    HCNonAdvectiveFreeComponentFlowBoundaryConditionData const& _data;
+    std::size_t const _local_matrix_size;
+    Eigen::RowVector3d const _surface_eigenvector;
```
Comment (TomFischer) [inline ProcessLib/BoundaryCondition/HCNonAdvectiveFreeComponentFlowBoundaryConditionLocalAssembler.h]:
```suggestion
    Eigen::RowVector3d const _surface_normal;
```

**Comment by jbathmann** [comment]
Thanks for your review. Just uploaded the changed commits.

**Comment by endJunction** [comment]
CTests look good now, there are some doxygen errors though: https://jenkins.opengeosys.org/job/ufz/job/ogs/job/PR-2466/14/doxygen/

**Comment by endJunction** [comment]
Nice!

**Comment by waltherm** [comment]
Nicely done. Two and a half questions:
1) Is the new keyword ```non_advective_form``` in the .prj mandatory? If not, what is the default value?
2) Is there a simple test for the documentation website for users to easily understand the difference?

**Comment by jbathmann** [comment]
@waltherm thanks for your questions. Here are the answers:
1.) The keyword is not mandatory. The default case is the "common" approach. This is documented here: https://doxygen.opengeosys.org/ogs_file_param__prj__processes__process__ComponentTransport__non_advective_form.html
2.) There is a new benchmark-test, but this is not prepared for the homepage, yet. The test ist called 2D_ComponentTransport_ImpermeableBoundaries and proves mass conservation, which is only possible using the non_advective_form.
